src/cares_wrap.cc

```diff
@@ -508,14 +511,14 @@ ChannelWrap::~ChannelWrap() {
   }
 
   ares_destroy(channel_);
-  CleanupTimer();
-}
 
+  if (timer_handle_ == nullptr)
+    return;
 
-void ChannelWrap::CleanupTimer() {
-  if (timer_handle_ == nullptr) return;
-
-  env()->CloseHandle(timer_handle_, [](uv_timer_t* handle) { delete handle; });
+  env()->SetUnrefImmediate([](Environment* env, void* data) {
+    env->CloseHandle(reinterpret_cast<uv_timer_t*>(data),
+                     [](uv_timer_t* handle) { delete handle; });
+  }, timer_handle_);
```

Comment:
So… what happens if we’re about to exit at this point? Do we just not close the handle? That creates a memory leak, i.e. I think we’d want to also close the handle from the `ChannelWrap` destructor.

Reply:
Correct me if I'm wrong but if it's too late to run the loop then we won't get to close it properly anyway? This is inside the destructor already... we're kind of at the mercy of the garbage collector timing.

Reply:
It’s not really too late to run the loop though, right? We’re just not allowing callbacks into JS anymore when we’re cleaning up, which makes the cleanup invisible from the JS PoV (and that is ultimately the goal here, right?)

Like, if I remember correctly this kind of situation is the reason that we’re running `CleanupHandles()` at the end of the `RunCleanup()` loop in `env.cc`

Reply:
Well, what we trying to avoid is emitting `beforeExit`, then having the GC run and resurrecting the loop again. As far as I can tell, `SetUnrefImmediate` works here except if the GC ran after the check stage during the `uv_run` inside the cleanup phase. I suppose we could check for `can_call_into_js` and either use Immediate or not depending on its state.

Reply:
Inside the cleanup phase we don’t really care about GC anymore I think, we’re destroying all C++ objects anyway?

> I suppose we could check for `can_call_into_js` and either use Immediate or not depending on its state.

What happens if we schedule it via `SetUnrefImmediate()` but never run that immediate because we don’t have anything else to clean up, so nothing refs the event loop?

Reply:
Maybe another question: Do we ever want `CloseHandle()` to resurrect the loop’s “lifeness”?

Reply:
^ Yes, that's what I'm looking at now. Let me update this with something slightly different and see what you think :)